ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/worldstate/WorldStateDownloader.java

```diff
@@ -0,0 +1,199 @@
+      final BlockHeader header,
+      final BigQueue<NodeDataRequest> pendingRequests,
+      final int hashCountPerRequest,
+      final int maxOutstandingRequests,
+      final LabelledMetric<OperationTimer> ethTasksTimer) {
+    this.ethContext = ethContext;
+    this.worldStateStorage = worldStateStorage;
+    this.header = header;
+    this.pendingRequests = pendingRequests;
+    this.hashCountPerRequest = hashCountPerRequest;
+    this.maxOutstandingRequests = maxOutstandingRequests;
+    this.ethTasksTimer = ethTasksTimer;
+    // TODO: construct an updater that will commit changes periodically as updates accumulate
+    this.worldStateStorageUpdater = worldStateStorage.updater();
+
+    Hash stateRoot = header.getStateRoot();
+    if (stateRoot.equals(MerklePatriciaTrie.EMPTY_TRIE_ROOT_HASH)) {
+      // If we're requesting data for an empty world state, we're already done
+      future = CompletableFuture.completedFuture(null);
+      status = Status.DONE;
+    } else {
+      pendingRequests.enqueue(NodeDataRequest.createAccountDataRequest(header.getStateRoot()));
+    }
+  }
+
+  // TODO: add a cancel method
+  public CompletableFuture<Void> run() {
+    synchronized (this) {
+      if (status == Status.DONE || status == Status.RUNNING) {
+        return future;
+      }
+      status = Status.RUNNING;
+      future = new CompletableFuture<>();
+    }
+
+    requestNodeData();
+    // TODO: Complete exceptionally on timeout / stalled download
+    return future;
+  }
+
+  private void requestNodeData() {
+    if (sendingRequests.compareAndSet(false, true)) {
+      while (shouldRequestNodeData()) {
+        Optional<EthPeer> maybePeer = ethContext.getEthPeers().idlePeer(header.getNumber());
+
+        if (!maybePeer.isPresent()) {
+          // If no peer is available, wait and try again
+          waitForNewPeer()
+              .whenComplete(
+                  (r, t) -> {
+                    requestNodeData();
+                  });
+          break;
+        } else {
+          EthPeer peer = maybePeer.get();
+
+          // Collect data to be requested
+          List<NodeDataRequest> toRequest = new ArrayList<>();
+          for (int i = 0; i < hashCountPerRequest; i++) {
+            if (pendingRequests.isEmpty()) {
+              break;
+            }
+            toRequest.add(pendingRequests.dequeue());
+          }
+
+          // Request and process node data
+          outstandingRequests.incrementAndGet();
+          sendAndProcessRequests(peer, toRequest)
+              .whenComplete(
+                  (res, error) -> {
+                    if (outstandingRequests.decrementAndGet() == 0 && pendingRequests.isEmpty()) {
+                      // We're done
+                      worldStateStorageUpdater.commit();
+                      future.complete(null);
+                    } else {
+                      // Send out additional requests
+                      requestNodeData();
```

Comment:
So it does. I wonder if it's worth extracting a method for the send single request function to make that while a bit easier to spot (though sometimes there's just no helping fools like me...).

Line 128 above:
`if (outstandingRequests.decrementAndGet() == 0 && pendingRequests.isEmpty()) {`
will then depend on `isEmpty()` being thread safe and completely accurate.  For the in memory implementation it would be enough to delegate to `ConccurrentLinkedQueue.isEmpty()` rather than using the separately tracked `size` variable I think.  We know we're the last request to finish so there's nothing else coming in at the same time, but we need to be sure that any pending requests added by the request prior to us have actually been reflected in the `pendingRequests` properly.